pertpy/tools/_mixscape.py

```diff
@@ -0,0 +1,225 @@
+                  the perturbation signature for every replicate separately.
+        pval_cutoff: P-value cut-off for selection of significantly DE genes.
+        fine_mode: When this is equal to TRUE, DE genes for each target gene class will be calculated for each gRNA separately
+                   and pooled into one DE list for calculating the perturbation score of every cell and their subsequent classification.
+        fine_mode_labels: The column of `.obs` with gRNA ID labels.
+        prtb_type: specify type of CRISPR perturbation expected for labeling mixscape classifications. Default is KO.
+        copy: Determines whether a copy of the `adata` is returned.
+        **kwargs: Additional arguments for the `NNDescent` class from `pynndescent`.
+
+    Returns:
+        If `copy=True`, returns the copy of `adata` with the classification result in `.obs`.
+        Otherwise writes the results directly to `.obs` of the provided `adata`.
+
+        mixscape_class: pandas.Series (`adata.obs['mixscape_class']`).
+        Classification result with cells being either classified as perturbed (KO, by default) or non-perturbed (NP) based on their target gene class.
+
+        mixscape_class_global: pandas.Series (`adata.obs['mixscape_class_global']`).
+        Global classification result (perturbed, NP or NT)
+
+        mixscape_class_p_ko: pandas.Series (`adata.obs['mixscape_class_p_ko']`).
+        Posterior probabilities used to determine if a cell is KO (default). Name of this item will change to match prtb.type parameter setting. (>0.5) or NP
+    """
+    if copy:
+        adata = adata.copy()
+
+    if split_by is None:
+        split_masks = [np.full(adata.n_obs, True, dtype=bool)]
+        categories = ["all"]
+    else:
+        split_obs = adata.obs[split_by]
+        categories = split_obs.unique()
+        split_masks = [split_obs == category for category in categories]
+
+    # determine gene sets across all splits/groups
+    perturbation_markers = {}
+    for split, split_mask in enumerate(split_masks):
+        category = categories[split]
+        genes = list(set(adata[split_mask].obs[labels]).difference([control]))
+        adata_split = adata[split_mask].copy()
+        sc.tl.rank_genes_groups(adata_split, layer=layer, groupby=labels, groups=genes, reference=control)
+        for gene in genes:
+            logfc_threshold_mask = adata_split.uns["rank_genes_groups"]["logfoldchanges"][gene] >= logfc_threshold
+            de_genes = adata_split.uns["rank_genes_groups"]["names"][gene][logfc_threshold_mask]
+            pvals_adj = adata_split.uns["rank_genes_groups"]["pvals_adj"][gene][logfc_threshold_mask]
+            de_genes = de_genes[pvals_adj < pval_cutoff]
+            if len(de_genes) < min_de_genes:
+                de_genes = np.array([])
+            perturbation_markers[(category, gene)] = de_genes
+
+    all_markers = [list(perturbation_markers[key]) for key in perturbation_markers.keys()][0]
```

Comment:
Could this block of code get a comment above it? What and why are we doing this? Just 1 sentence.